File: go-honeypot/server/internal/honeypots/mysql.go
package honeypots

import (
	"bytes"
	"context"
	cryptorand "crypto/rand"
	"encoding/binary"
	"encoding/hex"
	mrand "math/rand"
	"net"

	"github.com/h4ux/honeystack/go-honeypot/server/internal/config"
	"github.com/h4ux/honeystack/go-honeypot/server/internal/eventlog"
)

func NewMySQL(cfg config.Service, store *eventlog.Store) *TCP {
	var t *TCP
	t = NewTCP("mysql", cfg, store, func(ctx context.Context, conn net.Conn, sessionID string, meta ConnMeta) {
		cfg := t.Cfg()
		version := cfg.ServerVersion
		if version == "" {
			version = "8.0.36"
		}
		salt1 := make([]byte, 8)
		_, _ = cryptorand.Read(salt1)
		salt2 := make([]byte, 12)
		_, _ = cryptorand.Read(salt2)
		threadID := mrand.Uint32()

		var b bytes.Buffer
		b.WriteByte(0x0a)
		b.WriteString(version)
		b.WriteByte(0)
		_ = binary.Write(&b, binary.LittleEndian, threadID)
		b.Write(salt1)
		b.WriteByte(0)
		_ = binary.Write(&b, binary.LittleEndian, uint16(0xffff))
		b.WriteByte(0x21)
		_ = binary.Write(&b, binary.LittleEndian, uint16(0x0002))
		_ = binary.Write(&b, binary.LittleEndian, uint16(0x807f))
		b.WriteByte(0x15)
		b.Write(make([]byte, 10))
		b.Write(salt2)
		b.WriteByte(0)

		hdr := packetHeader(b.Len(), 0)
		_, _ = conn.Write(append(hdr, b.Bytes()...))

		received := make([]byte, 0, 1024)
		buf := make([]byte, 1024)
		for {
			n, err := conn.Read(buf)
			if err != nil {
				return
			}
			received = append(received, buf[:n]...)
			if len(received) > 8192 {
				return
			}
			if user, authHex, database, ok := parseHandshakeResponse(received); ok {
				pw := ""
				if authHex != "" {
					pw = "<sha1:" + authHex + ">"
				}
				accepted := shouldAccept(cfg.FakeAuth, user, pw)
				store.SetSessionUsername(sessionID, user)
				store.Log(eventlog.Event{
					Service: "mysql", Type: ternary(accepted, "auth_success", "login_attempt"), SessionID: sessionID,
					RemoteIP: meta.RemoteIP, RemotePort: meta.RemotePort, LocalPort: meta.LocalPort,
					Username: user, Password: pw,
					Details: map[string]any{"database": database, "accepted": accepted},
				})
				if !accepted {
					errPkt := errorPacket(1045, "28000", "Access denied for user '"+user+"'@'"+meta.RemoteIP+"' (using password: YES)")
					_, _ = conn.Write(errPkt)
					return
				}
				_, _ = conn.Write(okPacket(2))
				received = received[:0]
				continue
			}
			_ = ctx
		}
	})
	return t
}

func packetHeader(bodyLen, seq int) []byte {
	h := make([]byte, 4)
	h[0] = byte(bodyLen)
	h[1] = byte(bodyLen >> 8)
	h[2] = byte(bodyLen >> 16)
	h[3] = byte(seq)
	return h
}

func parseHandshakeResponse(buf []byte) (string, string, string, bool) {
	if len(buf) < 5 {
		return "", "", "", false
	}
	payloadLen := int(buf[0]) | int(buf[1])<<8 | int(buf[2])<<16
	if len(buf) < 4+payloadLen {
		return "", "", "", false
	}
	p := buf[4 : 4+payloadLen]
	if len(p) < 32 {
		return "", "", "", false
	}
	off := 32
	end := bytes.IndexByte(p[off:], 0)
	if end < 0 {
		return "", "", "", false
	}
	username := string(p[off : off+end])
	off += end + 1
	if off >= len(p) {
		return username, "", "", true
	}
	authLen := int(p[off])
	off++
	if off+authLen > len(p) {
		return username, "", "", true
	}
	auth := p[off : off+authLen]
	off += authLen
	var database string
	if off < len(p) {
		if idx := bytes.IndexByte(p[off:], 0); idx >= 0 {
			database = string(p[off : off+idx])
		}
	}
	authHex := hex.EncodeToString(auth)
	if len(authHex) > 40 {
		authHex = authHex[:40]
	}
	return username, authHex, database, true
}

func errorPacket(code int, sqlState, message string) []byte {
	var body bytes.Buffer
	body.WriteByte(0xff)
	_ = binary.Write(&body, binary.LittleEndian, uint16(code))
	body.WriteByte('#')
	body.WriteString(sqlState)
	body.WriteString(message)
	return append(packetHeader(body.Len(), 2), body.Bytes()...)
}

func okPacket(seq int) []byte {
	body := []byte{0x00, 0x00, 0x00, 0x02, 0x00, 0x00, 0x00}
	return append(packetHeader(len(body), seq), body...)
}

File: src/honeypots/mysql.js
'use strict';

const crypto = require('crypto');
const { TcpHoneypot } = require('./base');

class MysqlHoneypot extends TcpHoneypot {
  constructor(cfg, logger) {
    super({ name: 'mysql', port: cfg.port }, logger);
    this.cfg = cfg;
  }

  updateConfig(cfg) { this.cfg = cfg; }

  onConnection(socket, meta) {
    const version = (this.cfg.serverVersion || '8.0.36') + '\0';
    const threadId = Math.floor(Math.random() * 1e6);
    const salt1 = crypto.randomBytes(8);
    const salt2 = crypto.randomBytes(12);

    const capabilities = 0xffff;
    const capabilitiesUpper = 0x807f;
    const buf = Buffer.concat([
      Buffer.from([0x0a]),
      Buffer.from(version, 'binary'),
      u32le(threadId),
      salt1,
      Buffer.from([0x00]),
      u16le(capabilities & 0xffff),
      Buffer.from([0x21]),
      u16le(0x0002),
      u16le(capabilitiesUpper),
      Buffer.from([0x15]),
      Buffer.alloc(10, 0),
      salt2,
      Buffer.from([0x00])
    ]);
    const packet = Buffer.concat([lengthAndSeq(buf.length, 0), buf]);
    try { socket.write(packet); } catch { /* noop */ }

    let received = Buffer.alloc(0);
    socket.on('data', (data) => {
      received = Buffer.concat([received, data]).slice(0, 8192);
      const parsed = parseHandshakeResponse(received);
      if (parsed && parsed.username) {
        this.logger.log({
          service: 'mysql',
          type: 'login_attempt',
          sessionId: meta.sessionId,
          remoteIp: meta.remoteIp,
          remotePort: meta.remotePort,
          localPort: meta.localPort,
          username: parsed.username,
          password: parsed.authHex ? '<sha1:' + parsed.authHex + '>' : null,
          details: { database: parsed.database, plugin: parsed.plugin }
        });
        const err = buildErrorPacket(1045, "28000", `Access denied for user '${parsed.username}'@'${meta.remoteIp || 'unknown'}' (using password: YES)`);
        try { socket.write(err); } catch { /* noop */ }
        try { socket.end(); } catch { /* noop */ }
      }
    });
  }
}

function u16le(v) { const b = Buffer.alloc(2); b.writeUInt16LE(v, 0); return b; }
function u32le(v) { const b = Buffer.alloc(4); b.writeUInt32LE(v, 0); return b; }
function lengthAndSeq(len, seq) {
  const b = Buffer.alloc(4);
  b.writeUIntLE(len, 0, 3);
  b[3] = seq;
  return b;
}

function parseHandshakeResponse(buf) {
  if (buf.length < 5) return null;
  const payloadLen = buf.readUIntLE(0, 3);
  if (buf.length < 4 + payloadLen) return null;
  const p = buf.slice(4, 4 + payloadLen);
  if (p.length < 32) return null;
  // clientFlags(4) maxPacket(4) charset(1) reserved(23) username(null-term)
  let off = 32;
  const end = p.indexOf(0x00, off);
  if (end < 0) return null;
  const username = p.slice(off, end).toString('utf8');
  off = end + 1;
  if (off >= p.length) return { username };
  const authLen = p[off];
  off += 1;
  const auth = p.slice(off, off + authLen);
  off += authLen;
  let database = null;
  const dbEnd = p.indexOf(0x00, off);
  if (dbEnd >= 0) database = p.slice(off, dbEnd).toString('utf8');
  return { username, authHex: auth.toString('hex').slice(0, 40), database, plugin: null };
}

function buildErrorPacket(code, sqlState, message) {
  const body = Buffer.concat([
    Buffer.from([0xff]),
    u16le(code),
    Buffer.from('#'),
    Buffer.from(sqlState),
    Buffer.from(message)
  ]);
  return Buffer.concat([lengthAndSeq(body.length, 2), body]);
}

module.exports = { create: (cfg, logger) => new MysqlHoneypot(cfg, logger) };
